### PR TITLE
Flux editor

### PR DESCRIPTION
**Pull request by spatializes**
#### What's this PR do?

Moves the editor and keypress bindings into Alt Flux.
#### Where should the reviewer start?

Look at the new actions and stores for editor and keys. Look at the key-extensions to see how keypresses are now registered using Flux actions.
#### How should this be manually tested?

Test the editor to ensure that you can still type text, save files, etc.
Test the six keypresses that are present in the keys Flux Store. All should function as they did before.
#### Any background context you want to provide?

Bringing basic editor and keypress functionality into Flux using Alt.
#### What are the relevant tickets?

References #115

### CODE REVIEW

File: src/actions/keys.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+'use strict';
+
+const alt = require('../alt');
+
+class KeyActions {
+  findNext(evt) {
```
Comment (phated) [inline src/actions/keys.js]: move findNext and findPrevious into own action class

File: src/actions/keys.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    this.dispatch(evt);
+  }
+  findPrevious(evt) {
+    this.dispatch(evt);
+  }
+  moveByScrollUpLine(evt) {
```
Comment (phated) [inline src/actions/keys.js]: move moveByScrollUpLine and moveByScrollDownLine into own action class

File: src/actions/keys.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    this.dispatch(evt);
+  }
+  moveByScrollDownLine(evt) {
+    this.dispatch(evt);
+  }
+  tab(evt) {
```
Comment (phated) [inline src/actions/keys.js]: name this indent (?) and move to own action class (does it make sense to group with any other actions?)

File: src/actions/keys.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    this.dispatch(evt);
+  }
+  print(evt) {
```
Comment (phated) [inline src/actions/keys.js]: move to own action class

File: src/stores/keys.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      onFindPrevious: findPrevious,
+      onMoveByScrollUpLine: moveByScrollUpLine,
+      onMoveByScrollDownLine: moveByScrollDownLine,
+      onTab: tab,
+      onPrint: print
+    });
+
+  }
+
+  onFindNext(evt) {
```
Comment (phated) [inline src/stores/keys.js]: `evt` shouldn't be passed in here, it won't allow this to be used generically

File: src/stores/keys.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    const { cm } = this.getInstance();
+
+    evt.preventDefault();
+    cm.execCommand('findNext');
+  }
+  onFindPrevious(evt) {
```
Comment (phated) [inline src/stores/keys.js]: `evt` shouldn't be passed in here, it won't allow this to be used generically

File: src/stores/keys.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    const { cm } = this.getInstance();
+
+    evt.preventDefault();
+    cm.execCommand('findPrev');
+  }
+  onMoveByScrollUpLine(evt) {
```
Comment (phated) [inline src/stores/keys.js]: `evt` shouldn't be passed in here, it won't allow this to be used generically

File: src/stores/keys.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    const { cm } = this.getInstance();
+
+    evt.preventDefault();
+    const scrollbox = cm.getScrollInfo();
+    cm.scrollTo(null, scrollbox.top - cm.defaultTextHeight());
+  }
+  onMoveByScrollDownLine(evt) {
```
Comment (phated) [inline src/stores/keys.js]: `evt` shouldn't be passed in here, it won't allow this to be used generically

File: src/stores/keys.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    const { cm } = this.getInstance();
+
+    evt.preventDefault();
+    const scrollbox = cm.getScrollInfo();
+    cm.scrollTo(null, scrollbox.top + cm.defaultTextHeight());
+  }
+  onTab(evt) {
```
Comment (phated) [inline src/stores/keys.js]: `evt` shouldn't be passed in here, it won't allow this to be used generically

File: src/stores/keys.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    const { cm } = this.getInstance();
+
+    evt.preventDefault();
+    cm.execCommand('insertSoftTab');
+  }
+  onPrint(evt) {
```
Comment (phated) [inline src/stores/keys.js]: `evt` shouldn't be passed in here, it won't allow this to be used generically

File: plugins/editor/key-extension.js [hunk 11]
```diff
@@ -1,57 +1,37 @@
...
+  setup: function(app) {
 
     const cmCommands = {
       findNext: {
         code: 'F3',
-        exec: function(evt) {
-          evt.preventDefault();
-          cm.execCommand('findNext');
-        }
+        exec: findNext
```
Comment (phated) [inline plugins/editor/key-extension.js]: events should be handled in function wrappers here and not passed to the actions

File: plugins/editor/key-extension.js [hunk 12]
```diff
@@ -1,57 +1,37 @@
 'use strict';
 
-var CodeMirror = require('codemirror');
+var { findNext, findPrevious, moveByScrollUpLine,
```
Comment (phated) [inline plugins/editor/key-extension.js]: too many things destructured from 1 store, need to split these up

File: plugins/editor/index.js [hunk 13]
```diff
@@ -94,7 +96,8 @@ function editor(app, opts, done){
         'Ctrl-Down': false,
         Tab: false
       });
-      keyExtension.setup(app, codeEditor);
+      keyExtension.setup(app);
```
Comment (phated) [inline plugins/editor/index.js:66]: i need to think about this API some more, i'd prefer not having to call `.setup` for this but not sure how to get around it

File: src/actions/keys.js (hunk 3 above)
Comment (spatializes) [inline src/actions/keys.js]: onFindNext, onFindPrevious, onMoveByScollUpLine, onMoveByScrollDownLine, and Indent are all code mirror specific shortcuts - so i think it makes sense to group them all into one action class? called editorCommands maybe? then print would be its own class, called systemCommands?

Comment (phated) [inline src/actions/keys.js]: No, just because they all relate to the editor doesn't make them logically the same actions.

**Pull request merged**